Technibel support

Quentinbri opened the pull request:
Add Technibel A/C protocol
Modifications tested IRL on Technibel IRO 13 PLUS

Quentinbri committed:
Add Technibel support

Quentinbri committed:
Merge branch 'master' into Technibel_Support

# Conflicts:
#	src/IRac.cpp
#	src/IRrecv.cpp
#	src/IRrecv.h
#	src/IRremoteESP8266.h
#	src/IRsend.h
#	src/IRtext.cpp

Quentinbri committed:
Fix issues raised with C++ linter

crankyoldgit commented:
@Quentinbri First off, this is amazing. Congrats on doing this. It's a huge effort, I know.
I'll start a proper review of it soon, but I glanced over it on my phone and noticed some minor issues, and alas, one large one.

The "large" issue is I'm pretty sure you've chosen the wrong bit-order for the message. If you choose MSBF order (instead of the LSBF order you've picked), you shouldn't need to do all the bit reversing for the numeric values. e.g. Temperature.
While it will be a lot of work to undo that decision, it will reduce the complexity and the size of the code, and also increase the speed of it.
If you don't feel up to doing that, I can allow it as is, but I'll change it straight away. i.e. The decoded values etc will all be changed/reversed. Your choice. As you've got access to the device, I'd prefer it if you did, that way you can more easily verify the process as you go.

Anyway, again, this is a TOP effort. Thanks very very much for doing it and contributing.

crankyoldgit commented on examples/TurnOnTechnibelAC/TurnOnTechnibelAC.ino:
```diff
@@ -0,0 +1,69 @@
+/* Copyright 2017, 2018 David Conran
```
Please remove the example entirely. We already have numerous `TurnOnBlahAC.ino` examples, of which this is pretty similar. Adding it means our build times for the libraries gets increased for not much real value.

crankyoldgit commented on src/IRac.cpp:
```diff
@@ -251,6 +252,9 @@ bool IRac::isProtocolSupported(const decode_type_t protocol) {
 #if SEND_TCL112AC
     case decode_type_t::TCL112AC:
 #endif
+#if SEND_TECO
```
Cut & paste error, I think you mean:

`#if SEND_TECHNIBEL_AC`

crankyoldgit commented on src/IRac.cpp:
```diff
@@ -3177,6 +3181,13 @@ namespace IRAcUtils {
 #endif  // DECODE_VESTEL_AC
+#if DECODE_TECHNIBEL_AC
+      case decode_type_t::TECHNIBEL_AC: {
+        IRTechnibelAc ac(0);
```
Instead, please use:
`        IRTechnibelAc ac(kGpioUnused);`

crankyoldgit commented on src/IRrecv.h:
```diff
@@ -663,6 +663,12 @@ class IRrecv {
+  bool decodeTechnibelAc(decode_results *results,
+                     uint16_t offset = kStartOffset,
+                     const uint16_t nbits = kTechnibelAcBits,
+                     const bool strict = true);
```
Style nitpick: Please align these lines to start under the `d` of decode_results. i.e. after the `(` column.

crankyoldgit commented on src/IRsend.h:
```diff
@@ -631,6 +631,10 @@ class IRsend {
 #endif
+#if SEND_TECHNIBEL_AC
+  void sendTechnibelAc(uint64_t data, uint16_t nbits = kTechnibelAcBits,
+                      uint16_t repeat = kTechnibelAcDefaultRepeat);
```
Minor style nitpick again: Please start the line so it aligns with the `u` in `uint64_t`. i.e. column after the `(`.

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -0,0 +1,416 @@
+//           Typically kTechnibelAcBits.
+//   repeat: The number of times the command is to be repeated.
+//
+// Status: STABLE / Reported as working on a real device.
```
Please replace with `doxygen` style comments as per all other functions/procedures etc.
e.g.
```
/// Send an Technibel AC formatted message.
/// Status: STABLE / Reported as working on a real device.
/// @param[in] data containing the IR command.
/// @param[in] nbits Nr. of bits to send. usually kTechnibelAcBits
/// @param[in] repeat Nr. of times the message is to be repeated.
```

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -0,0 +1,416 @@
+              kTechnibelAcBitMark, kTechnibelAcOneSpace,
+              kTechnibelAcBitMark, kTechnibelAcZeroSpace,
+              kTechnibelAcBitMark, kTechnibelAcGap,
+              data, nbits, kTechnibelAcFreq, false,  // LSB First.
```
As mentioned elsewhere, using MSB First will negate having to use `reverseBits()` elsewhere.
e.g.
`              data, nbits, kTechnibelAcFreq, true,  // MSB First.`

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -0,0 +1,416 @@
+// Returns:
+//   boolean: True if it can decode it, false if it can't.
+//
+// Status: STABLE / Expected to be working.
```
Ditto earlier doxygen-style comments comment.
```
/// Decode the supplied Technibel A/C message.
/// Status: STABLE / Expected to be working
/// @param[in,out] results Ptr to the data to decode & where to store the decode
/// @param[in] offset The starting index to use when attempting to decode the
///   raw data. Typically/Defaults to kStartOffset.
/// @param[in] nbits The number of data bits to expect. Typically kTechnibelAcBits.
/// @param[in] strict Flag indicating if we should perform strict matching.
/// @return A boolean. True if it can decode it, false if it can't.
```

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -0,0 +1,416 @@
+  this->setTemp(_saved_temp);
+  this->setTempUnit(_saved_temp_units);
+  this->setMode(kTechnibelAcCool);
+  this->setFan(kTechnibelAcFanLow);
```
Please drop the use of `this->` unless it is really needed anywhere in the code. It's a style decision we came up with only recently. We are slowly removing it from all the existing code. So might as well not add more instances of it.

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -0,0 +1,416 @@
+    case stdAc::swingv_t::kOff:
+      return false;
+    case stdAc::swingv_t::kAuto:
+      return true;
```
This clause is redundant. If you remove it, you'll get the same result due to the `default:` clause.

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -0,0 +1,416 @@
+  return GETBIT64(remote_state, kTechnibelAcTimerEnableBit);
+}
+
+// Set the timer to shutdown the device in nr of minutes.
```
Comment is inconsistent with the argument.

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -0,0 +1,416 @@
+// Set the timer to shutdown the device in nr of minutes.
+// Args:
+//   nr_of_hours: Total nr of hours to wait before shutting down the device.
+//               (Max 24 hrs)
```
If possible, please use `nr_of_minutes` to keep it consistent with the rest of the protocols.
i.e. Just do a `/ 60` in the function procedure.

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -0,0 +1,416 @@
+uint8_t IRTechnibelAc::getTimer(void) {
+  uint8_t timer = GETBITS64(remote_state, kTechnibelAcTimerHoursOffset,
+                            kTechnibelAcHoursSize);
+  return reverseBits(timer, kTechnibelAcHoursSize);
```
Similarly, please return the `nr_of_mins` instead of hours, to be consistent with other A/C protocols.
i.e.
`return reverseBits(timer, kTechnibelAcHoursSize) * 60;`

crankyoldgit commented on src/ir_Technibel.h:
```diff
@@ -0,0 +1,151 @@
+// Ref:
+//
+
+// Kudos:
+//   : For the breakdown and mapping of the bit values.
```
As you wrote it, and your names in the copyright, you can probably remove these comments. ;-)

crankyoldgit commented on test/ir_Technibel_test.cpp:
```diff
@@ -0,0 +1,277 @@
+// Copyright 2020 David Conran
```
Add your name here. ;-)

crankyoldgit commented on test/ir_Technibel_test.cpp:
```diff
@@ -0,0 +1,277 @@
+      "Sleep: Off, Swing(V): On, Timer: Off",
+      IRAcUtils::resultAcToString(&irsend.capture));
+  // stdAc::state_t r, p;
+  // ASSERT_TRUE(IRAcUtils::decodeToState(&irsend.capture, &r, &p));
```
This should work normally. Is there a reason it's commented out?

crankyoldgit reviewed (changes requested):
You're also missing a section in:
`IRac::decodeToState()` in IRac.cpp. (I can add this later)
an `IRac::technibel()` function in IRac.cpp (I can add this later)
and a corresponding call to `technibel()` in `bool IRac::sendAc(const stdAc::state_t desired, const stdAc::state_t *prev)` function. (I can add this later)

Sorry for all the nitpicks etc, everything should be fairly minor, except for the LSB->MSB swap. that will be a doosey :-(

Again, top effort. This is amazing work. Please keep it up!

Quentinbri commented:
Thank you so much for your comments @crankyoldgit.
I will handle all these issues by myself.

Quentinbri committed:
Change bit-order of the message from LSBF to MSBF
Change comments into doxygen ones
Update a section of IRac.cpp
Remove TurnOnTechnibelAC example
Various minor improvements

Quentinbri committed:
Fix Doxygen issue in IRac.cpp

crankyoldgit commented on src/ir_Technibel.cpp:
```diff
@@ -30,41 +30,32 @@ const uint16_t kTechnibelAcOverhead = 3;
               kTechnibelAcBitMark, kTechnibelAcZeroSpace,
               kTechnibelAcBitMark, kTechnibelAcGap,
-              data, nbits, kTechnibelAcFreq, false,  // LSB First.
+              data, nbits, kTechnibelAcFreq, true,  // LSB First.
```
Comment is now incorrect. LSB -> MSB

crankyoldgit reviewed (approved):
Looks good to me. Thanks for all the improvements and the code itself!

I'll merge this and make some minor improvements/changes/tests.

crankyoldgit commented:
FYI, the changes mentioned above have now been included in the new v2.7.11 release of the library.
